// @shelby-protocol/sdk/dist/chunk-SEXQTDX6.mjs
// src/core/constants.ts
import { Network } from "@aptos-labs/ts-sdk";
var DEFAULT_PROJECT_NAME = "shelby";
var DEFAULT_PROJECT_DESCRIPTION = "High performance, decentralized storage";
var NetworkToShelbyRPCBaseUrl = {
  [Network.SHELBYNET]: "https://api.shelbynet.shelby.xyz/shelby",
  [Network.DEVNET]: void 0,
  [Network.TESTNET]: void 0,
  [Network.MAINNET]: void 0,
  [Network.LOCAL]: void 0,
  [Network.CUSTOM]: void 0
};
var NetworkToShelbyBlobIndexerBaseUrl = {
  [Network.SHELBYNET]: "https://api.shelbynet.aptoslabs.com/nocode/v1/public/cmforrguw0042s601fn71f9l2/v1/graphql",
  [Network.DEVNET]: void 0,
  [Network.TESTNET]: void 0,
  [Network.MAINNET]: void 0,
  [Network.LOCAL]: void 0,
  [Network.CUSTOM]: void 0
};
var SHELBY_DEPLOYER = "0xc63d6a5efb0080a6029403131715bd4971e1149f7cc099aac69bb0069b3ddbf5";
var TOKEN_DEPLOYER = "0x33009e852be7f93762dd0bf303383c2cb2c5cab7a30d8238ca5f9f177ae75124";
var TOKEN_OBJECT_ADDRESS = "0x249f5c642a63885ff88a5113b3ba0079840af5a1357706f8c7f3bfc5dd12511f";
var SHELBYUSD_TOKEN_ADDRESS = TOKEN_OBJECT_ADDRESS;
var SHELBYUSD_TOKEN_MODULE = "shelby_usd";
var SHELBYUSD_TOKEN_NAME = "ShelbyUSD";
var SHELBYUSD_FA_METADATA_ADDRESS = "0x1b18363a9f1fe5e6ebf247daba5cc1c18052bb232efdc4c50f556053922d98e1";

export {
  DEFAULT_PROJECT_NAME,
  DEFAULT_PROJECT_DESCRIPTION,
  NetworkToShelbyRPCBaseUrl,
  NetworkToShelbyBlobIndexerBaseUrl,
  SHELBY_DEPLOYER,
  TOKEN_DEPLOYER,
  TOKEN_OBJECT_ADDRESS,
  SHELBYUSD_TOKEN_ADDRESS,
  SHELBYUSD_TOKEN_MODULE,
  SHELBYUSD_TOKEN_NAME,
  SHELBYUSD_FA_METADATA_ADDRESS
};
