// @shelby-protocol/sdk/dist/chunk-ZPW742E7.mjs
// src/core/erasure/constants.ts
var ErasureCodingScheme = /* @__PURE__ */ ((ErasureCodingScheme2) => {
  ErasureCodingScheme2["ClayCode_16Total_10Data_13Helper"] = "ClayCode_16Total_10Data_13Helper";
  return ErasureCodingScheme2;
})(ErasureCodingScheme || {});
var ERASURE_CODE_PARAMS = {
  ["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */]: {
    // total chunks (data + parity)
    erasure_n: 16,
    // data chunks
    erasure_k: 10,
    // helper nodes
    erasure_d: 13
  }
};
var DEFAULT_ERASURE_N = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_n;
var DEFAULT_ERASURE_K = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_k;
var DEFAULT_ERASURE_D = ERASURE_CODE_PARAMS["ClayCode_16Total_10Data_13Helper" /* ClayCode_16Total_10Data_13Helper */].erasure_d;
var DEFAULT_ERASURE_M = DEFAULT_ERASURE_N - DEFAULT_ERASURE_K;

export {
  ErasureCodingScheme,
  ERASURE_CODE_PARAMS,
  DEFAULT_ERASURE_N,
  DEFAULT_ERASURE_K,
  DEFAULT_ERASURE_D,
  DEFAULT_ERASURE_M
};
